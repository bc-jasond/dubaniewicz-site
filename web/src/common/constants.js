export const PAGE_NAME_VIEW = 'public';
export const PAGE_NAME_EDIT = 'edit';
export const PAGE_NAME_PUBLIC = 'public';
export const PAGE_NAME_PRIVATE = 'private';
export const PAGE_NAME_MANAGE = 'manage';
export const PAGE_NAME_USER_PROFILE = 'user';

// EDITOR
export const NEW_POST_URL_ID = 'new';
// TODO: this is a placeholder to be able to set the caret in an empty tag
export const ZERO_LENGTH_CHAR = '\u200B';
// key codes from a 2015 macbook 13" & Kinesis Advantage2
// TODO: keyCode is deprecated
export const KEYCODE_BACKSPACE = 8;
export const KEYCODE_TAB = 9;
export const KEYCODE_ENTER = 13;
export const KEYCODE_SHIFT_RIGHT = 16;
export const KEYCODE_CTRL = 17;
export const KEYCODE_ALT = 18;
export const KEYCODE_CAPS_LOCK = 20;
export const KEYCODE_ESC = 27;
export const KEYCODE_SPACE = 32;
export const KEYCODE_PAGE_UP = 33;
export const KEYCODE_PAGE_DOWN = 34;
export const KEYCODE_END = 35;
export const KEYCODE_HOME = 36;
export const KEYCODE_LEFT_ARROW = 37;
export const KEYCODE_UP_ARROW = 38;
export const KEYCODE_RIGHT_ARROW = 39;
export const KEYCODE_DOWN_ARROW = 40;
export const KEYCODE_DEL = 46;
export const KEYCODE_V = 86;
export const KEYCODE_X = 88;
export const KEYCODE_Z = 90;
export const KEYCODE_Y = 89;
export const KEYCODE_SHIFT_OR_COMMAND_LEFT = 91;
export const KEYCODE_COMMAND_RIGHT = 93;
export const KEYCODE_F1 = 112;
export const KEYCODE_F2 = 113;
export const KEYCODE_F3 = 114;
export const KEYCODE_F4 = 115;
export const KEYCODE_F5 = 116;
export const KEYCODE_F6 = 117;
export const KEYCODE_F7 = 118;
export const KEYCODE_F8 = 119;
export const KEYCODE_F9 = 120;
export const KEYCODE_F10 = 121;
export const KEYCODE_F11 = 122;
export const KEYCODE_F12 = 123;
export const KEYCODE_PRINT_SCREEN = 124;
export const KEYCODE_SPACE_NBSP = 160;

// post level edits
export const POST_ACTION_REDIRECT_TIMEOUT = 1000;

// ENV
export const API_URL =
  process.env.NODE_ENV === 'production'
    ? 'https://api.filbert.xyz'
    : 'http://localhost:3001';

// AUTH
export const FILBERT_LOCALSTORAGE_NAMESPACE = 'filbert';
export const AUTH_TOKEN_KEY = 'token';
export const SESSION_KEY = 'session';
export const SESSION_THEME = 'theme';
export const SESSION_FONT = 'font';

// DOM
export const DOM_ELEMENT_NODE_TYPE_ID = 1;
export const DOM_TEXT_NODE_TYPE_ID = 3;
export const DOM_INPUT_TAG_NAME = 'INPUT';


// THEME
export const DARK_MODE_THEME = 'dark';
export const LIGHT_MODE_THEME = 'light';
export const SANS_FONT_THEME = 'sans';
export const MIXED_FONT_THEME = 'mixed';
