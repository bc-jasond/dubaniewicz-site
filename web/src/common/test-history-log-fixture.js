export const nodeUpdatesByNodeId = [
  {
    post_id: 277,
    type: 'quote',
    content: '',
    meta: {
      quote: "Here's something to say",
      author: 'B Guys and Guy',
      context: 'some co',
    },
    id: '0fbe',
  },
  {
    post_id: 277,
    type: 'quote',
    content: '',
    meta: {
      quote: "Here's something to say",
      author: 'B Guys and Guy',
      context: 'some context h',
    },
    id: '0fbe',
  },
  {
    post_id: 277,
    type: 'quote',
    content: '',
    meta: {
      quote: "Here's something to say",
      author: 'B Guys and Guy',
      context: 'some context here',
      url: 'http://',
    },
    id: '0fbe',
  },
  {
    post_id: 277,
    type: 'quote',
    content: '',
    meta: {
      quote: "Here's something to say",
      author: 'B Guys and Guy',
      context: 'some context here',
      url: 'http://www.',
    },
    id: '0fbe',
  },
  {
    post_id: 277,
    type: 'quote',
    content: '',
    meta: {
      quote: "Here's something to say",
      author: 'B Guys and Guy',
      context: 'some context here',
      url: 'http://www.stuff.c',
    },
    id: '0fbe',
  },
  {
    post_id: 277,
    type: 'quote',
    content: '',
    meta: {
      quote: "Here's something to say",
      author: 'B Guys and Guy',
      context: 'some context here',
      url: 'http://www.stuff.com',
    },
    id: '0fbe',
  },
  {
    post_id: 277,
    type: 'quote',
    content: '',
    meta: {
      quote: "Here's something to say",
      author: 'B Guys and Guy',
      context: 'some context here',
      url: 'http://www.stuff.com',
    },
    id: '0fbe',
    next_sibling_id: 'a43f',
  },
  { post_id: 277, type: 'p', content: '', meta: {}, id: 'a43f' },
  { post_id: 277, type: 'p', content: 'And ano', meta: {}, id: 'a43f' },
  { post_id: 277, type: 'p', content: 'And another?', meta: {}, id: 'a43f' },
  {
    post_id: 277,
    type: 'p',
    content: 'And another?',
    meta: {},
    id: 'a43f',
    next_sibling_id: '1744',
  },
  { post_id: 277, type: 'p', content: '', meta: {}, id: '1744' },
  {
    post_id: 277,
    type: 'p',
    content: 'And another?',
    meta: {
      selections: {
        length: -1,
        'selection-bold': false,
        'selection-italic': false,
        'selection-code': false,
        'selection-siteinfo': false,
        'selection-mini': false,
        'selection-strikethrough': false,
        'selection-link': false,
        linkUrl: '',
      },
    },
    id: 'a43f',
    next_sibling_id: '1744',
  },
  '1744',
];
export const contentNodeHistoryLog = [
  {
    executeSelectionOffsets: { startNodeId: '0fbe' },
    unexecuteSelectionOffsets: { startNodeId: '0fbe' },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
          },
          id: '0fbe',
        },
        executeState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some co',
          },
          id: '0fbe',
        },
      },
    ],
  },
  {
    executeSelectionOffsets: { startNodeId: '0fbe' },
    unexecuteSelectionOffsets: { startNodeId: '0fbe' },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some co',
          },
          id: '0fbe',
        },
        executeState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context h',
          },
          id: '0fbe',
        },
      },
    ],
  },
  {
    executeSelectionOffsets: { startNodeId: '0fbe' },
    unexecuteSelectionOffsets: { startNodeId: '0fbe' },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context h',
          },
          id: '0fbe',
        },
        executeState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context here',
            url: 'http://',
          },
          id: '0fbe',
        },
      },
    ],
  },
  {
    executeSelectionOffsets: { startNodeId: '0fbe' },
    unexecuteSelectionOffsets: { startNodeId: '0fbe' },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context here',
            url: 'http://',
          },
          id: '0fbe',
        },
        executeState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context here',
            url: 'http://www.',
          },
          id: '0fbe',
        },
      },
    ],
  },
  {
    executeSelectionOffsets: { startNodeId: '0fbe' },
    unexecuteSelectionOffsets: { startNodeId: '0fbe' },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context here',
            url: 'http://www.',
          },
          id: '0fbe',
        },
        executeState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context here',
            url: 'http://www.stuff.c',
          },
          id: '0fbe',
        },
      },
    ],
  },
  {
    executeSelectionOffsets: { startNodeId: '0fbe' },
    unexecuteSelectionOffsets: { startNodeId: '0fbe' },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context here',
            url: 'http://www.stuff.c',
          },
          id: '0fbe',
        },
        executeState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context here',
            url: 'http://www.stuff.com',
          },
          id: '0fbe',
        },
      },
    ],
  },
  {
    executeSelectionOffsets: { startNodeId: 'a43f', caretStart: 0 },
    unexecuteSelectionOffsets: {
      startNodeId: '0fbe',
      caretStart: 0,
      caretEnd: 0,
    },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context here',
            url: 'http://www.stuff.com',
          },
          id: '0fbe',
        },
        executeState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context here',
            url: 'http://www.stuff.com',
          },
          id: '0fbe',
          next_sibling_id: 'a43f',
        },
      },
      {
        executeState: {
          post_id: 277,
          type: 'p',
          content: '',
          meta: {},
          id: 'a43f',
        },
      },
    ],
  },
  {
    executeSelectionOffsets: { startNodeId: 'a43f', caretStart: 7 },
    unexecuteSelectionOffsets: {
      caretStart: 0,
      caretEnd: 0,
      startNodeId: 'a43f',
    },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'p',
          content: '',
          meta: {},
          id: 'a43f',
        },
        executeState: {
          post_id: 277,
          type: 'p',
          content: 'And ano',
          meta: {},
          id: 'a43f',
        },
      },
    ],
  },
  {
    executeSelectionOffsets: { startNodeId: 'a43f', caretStart: 12 },
    unexecuteSelectionOffsets: {
      caretStart: 7,
      caretEnd: 7,
      startNodeId: 'a43f',
    },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'p',
          content: 'And ano',
          meta: {},
          id: 'a43f',
        },
        executeState: {
          post_id: 277,
          type: 'p',
          content: 'And another?',
          meta: {},
          id: 'a43f',
        },
      },
    ],
  },
  {
    executeSelectionOffsets: { startNodeId: '1744', caretStart: 0 },
    unexecuteSelectionOffsets: {
      caretStart: 12,
      caretEnd: 12,
      startNodeId: 'a43f',
    },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'p',
          content: 'And another?',
          meta: {},
          id: 'a43f',
        },
        executeState: {
          post_id: 277,
          type: 'p',
          content: 'And another?',
          meta: {},
          id: 'a43f',
          next_sibling_id: '1744',
        },
      },
      {
        executeState: {
          post_id: 277,
          type: 'p',
          content: '',
          meta: {},
          id: '1744',
        },
      },
      {
        unexecuteState: {
          post_id: 277,
          type: 'p',
          content: 'And another?',
          meta: {},
          id: 'a43f',
          next_sibling_id: '1744',
        },
        executeState: {
          post_id: 277,
          type: 'p',
          content: 'And another?',
          meta: {
            selections: {
              length: -1,
              'selection-bold': false,
              'selection-italic': false,
              'selection-code': false,
              'selection-siteinfo': false,
              'selection-mini': false,
              'selection-strikethrough': false,
              'selection-link': false,
              linkUrl: '',
            },
          },
          id: 'a43f',
          next_sibling_id: '1744',
        },
      },
      {
        unexecuteState: {
          post_id: 277,
          type: 'p',
          content: '',
          meta: {},
          id: '1744',
        },
      },
    ],
  },
];
