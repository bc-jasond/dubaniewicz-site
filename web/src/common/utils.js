import { Map } from 'immutable';
import { s4 } from '@filbert/util';

export function confirmPromise(msg) {
  return new Promise((resolve) => {
    if (confirm(msg)) {
      resolve(true);
      return;
    }
    resolve(false);
  });
}

export function formatPostDate(dateStr) {
  if (!dateStr) {
    return;
  }
  const publishedDate = new Date(dateStr);
  return publishedDate.toLocaleDateString('en-us', {
    // weekday: 'long',
    year: 'numeric',
    day: 'numeric',
    month: 'long',
  });
}

export function formatStreakDate(dateInt) {
  const date = new Date();
  date.setFullYear(parseInt(`${dateInt}`.substring(0, 4), 10));
  date.setMonth(parseInt(`${dateInt}`.substring(4, 6), 10));
  date.setDate(parseInt(`${dateInt}`.substring(6), 10));
  return date.toLocaleDateString('en-us', {
    // weekday: 'long',
    year: 'numeric',
    day: 'numeric',
    month: 'long',
  });
}

export function formatNumber(number) {
  return new Intl.NumberFormat().format(number);
}

export function getCharFromEvent(evt) {
  // Firefox issue when typing fast - skip these.  NOTE: this could ignore composing characters for languages other than English
  //https://stackoverflow.com/a/25509350/1991322
  if (evt.key === 'Process') {
    return '';
  }
  if (evt && typeof evt.keyCode !== 'undefined') {
    // for normal "found on the keyboard" characters
    return evt.key;
  }
  // for OS X emoji keyboard insert
  return evt.data;
}

export function getCanonicalFromTitle(title) {
  let canonical = title
    // take first 25 chars
    .substring(0, 25)
    .toLowerCase()
    // remove all whitespace, replace with hyphen
    .replace(/\s+/g, '-')
    // keep only alpha numeric chars
    .replace(/[^0-9a-z-]/g, '');
  // append a hash
  canonical += `-${s4()}`;
  // dedupe hyphens
  return canonical.replace(/-+/g, '-');
}

export function stopAndPrevent(evt) {
  if (evt && evt.stopPropagation && evt.preventDefault) {
    evt.stopPropagation();
    evt.preventDefault();
  }
}

export function usernameIsValid(maybeUsername) {
  return /^@[0-9a-z]{1,42}$/.test(maybeUsername);
}

export function idIsValid(maybeId) {
  return new RegExp(/[0-9a-f]{4}/).test(maybeId);
}

export function nodeIsValid(node) {
  return Map.isMap(node) && idIsValid(node.get('id'));
}

export function isBrowser() {
  return typeof window !== 'undefined';
}

export function getUrl(s) {
  try {
    const url = new URL(s);
    const urlString = url.toString();
    // URL().toString() adds the trailing / slash to the host portion of the url
    // this is annoying when you're typing
    // do a check to see if the end of the user input is a `/` and if not, strip it
    if (urlString.slice(-1) === '/' && s.slice(-1) !== '/') {
      return urlString.slice(0, -1);
    }
    return urlString;
  } catch (err) {
    return '';
  }
}
