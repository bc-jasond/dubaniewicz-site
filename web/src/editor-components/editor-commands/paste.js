import { assertValidDomSelectionOrThrow } from '../../common/dom.mjs';
import { NODE_CONTENT, NODE_TYPE } from '@filbert/document';

export function doPaste(documentModel, selectionOffsets, clipboardData) {
  assertValidDomSelectionOrThrow(selectionOffsets);
  const { caretStart, startNodeId } = selectionOffsets;

  // split selectedNodeContent at caret
  const clipboardText = clipboardData.getData('text/plain');
  // TODO: convert HTML => filbert documentModel
  // const clipboardHtml = clipboardData.getData('text/html');
  const startNode = documentModel.getNode(startNodeId);
  if (!startNode.isTextType()) {
    console.warn(
      `doPaste() - trying to paste into a "${startNode.type}" node is not supported`,
      selectionOffsets,
      clipboardData
    );
    return {};
  }
  const historyState = [];

  // TODO: preserve format of copied text
  const beforeContent = startNode.content;
  const contentLeft = startNode.content.substring(0, caretStart);
  const contentRight = startNode.content.substring(caretStart);
  console.info(
    'PASTE - paragraph content: ',
    contentLeft,
    contentRight,
    caretStart,
    clipboardText
  );
  const clipboardLines = clipboardText.split('\n');
  // single line paste
  if (clipboardLines.length === 1) {
    const updatedContent = `${contentLeft}${clipboardText}${contentRight}`;
    const { length: diffLength } = clipboardText;
    startNode.content = updatedContent;
    startNode.formatSelections.adjustSelectionOffsetsAndCleanup(
      updatedContent.length,
      beforeContent.length,
      caretStart,
      diffLength
    );
    historyState.push(documentModel.update(startNode));
    return {
      selectionOffsets: {
        startNodeId,
        caretStart: contentLeft.length + diffLength,
      },
      historyState,
    };
  }
  // MULTI LINE PASTE
  // NOTE: the order of operations is important here
  console.info('PASTE: multi-line');
  const firstLine = clipboardLines.shift();
  const lastLine = clipboardLines.pop();
  const leftNodeId = startNodeId;
  // add 'lastLine' content to beginning of "right"
  const updatedRightNodeContent = `${lastLine}${contentRight}`;
  // add 'firstLine' content to end of "left"
  const updatedLeftNodeContent = `${contentLeft}${firstLine}`;
  // TODO: this is largely copied from handleEnterTextType() above, maybe split the shared code into a helper?
  // NOTE: rightNode insert (last before first) so that leftNode gets updated with a next_sibling_id
  // 1) insert a new node (rightNode) for 'lastLine' - using BEFORE content
  historyState.push(
    documentModel.insertAfter(
      { [NODE_TYPE]: startNode.type, [NODE_CONTENT]: beforeContent },
      leftNodeId
    )
  );
  const rightNodeId = documentModel.lastInsertId;
  // now leftNode has 'next_sibling_id' set to rightNode
  // important: 'content' is now contentLeft
  const leftNode = documentModel.getNode(leftNodeId);
  leftNode.content = contentLeft;
  const rightNode = documentModel.getNode(rightNodeId);

  // 2) if the original selected node can have Selections - move them to the right node if needed
  // NOTE: do this with BEFORE content
  const {
    left,
    right,
  } = leftNode.formatSelections.splitSelectionsAtCaretOffset(caretStart);
  leftNode.formatSelections = left;
  rightNode.formatSelections = right;

  // 3) update content with firstLine & lastLine
  leftNode.content = updatedLeftNodeContent;
  rightNode.content = updatedRightNodeContent;

  // 4) adjust offsets with updated content
  leftNode.formatSelections.adjustSelectionOffsetsAndCleanup(
    updatedLeftNodeContent.length,
    contentLeft.length, // this is before content! takes this as an argument for comparison with now updated content in leftNode
    caretStart,
    firstLine.length
  );
  rightNode.formatSelections.adjustSelectionOffsetsAndCleanup(
    updatedRightNodeContent.length,
    contentRight.length, // before content!
    0,
    lastLine.length
  );
  historyState.push(documentModel.update(leftNode));

  // there are middle lines, insert Paragraphs after "left"
  let prevNodeId = startNodeId;
  while (clipboardLines.length > 0) {
    const currentLine = clipboardLines.shift();
    // skip whitespace only lines TODO: allow in Code sections?
    if (currentLine.trim().length > 0) {
      historyState.push(
        documentModel.insertAfter(
          { [NODE_TYPE]: leftNode.type, [NODE_CONTENT]: currentLine },
          prevNodeId
        )
      );
      const nextId = documentModel.lastInsertId;
      prevNodeId = nextId;
    }
  }

  historyState.push(documentModel.update(rightNode));
  return {
    selectionOffsets: {
      startNodeId: rightNodeId,
      caretStart: lastLine.length,
    },
    historyState,
  };
}
