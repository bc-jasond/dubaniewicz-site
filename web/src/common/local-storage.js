import { fromJS, isImmutable, Map } from 'immutable';
import { FILBERT_LOCALSTORAGE_NAMESPACE } from './constants';

// reading & parsing JSON from localStorage is too slow.
// take a dual-write approach - one sync to memory and one optimistic async to localStorage
let inMemoryCache = Map();
let writeTimeout = {};

// TODO: this is a cheap way out, this file should be mocked in consumer tests
export function clearForTests() {
  inMemoryCache = Map();
  writeTimeout = {};
}

function getNamespacedKey(key) {
  return `${FILBERT_LOCALSTORAGE_NAMESPACE}-${key}`;
}

export function set(key, value, debounce = true) {
  if (!value) {
    inMemoryCache = inMemoryCache.delete(key);
  } else {
    inMemoryCache = inMemoryCache.set(key, fromJS(value));
  }

  if (writeTimeout[key]) {
    clearTimeout(writeTimeout[key]);
  }
  const setLocalStorage = (k) => {
    try {
      const namespacedKey = getNamespacedKey(k);
      let item = inMemoryCache.get(k);
      if (isImmutable(item)) {
        item = item.toJS();
      }
      if (!item) {
        localStorage.removeItem(namespacedKey);
      } else {
        localStorage.setItem(namespacedKey, JSON.stringify(item));
      }
    } catch (err) {
      // sometimes the localStorage.setItem() call fails - probably exceeded quota
      console.error(err);
      if (err?.name === 'QuotaExceededError') {
        // truncate history
      }
    }
  };
  if (!debounce) {
    setLocalStorage(key);
    return;
  }
  // this is a "backup" of the in memory storage to localStorage (for page reload in the same browser)
  // TODO: this could be written to a central storage like the DB or redis
  // this is a slow operation, only do it every so often
  writeTimeout[key] = setTimeout(setLocalStorage.bind(null, key), 3000);
}

export function get(key, defaultValue) {
  const namespacedKey = getNamespacedKey(key);
  try {
    if (!inMemoryCache.has(key)) {
      const currentLocal = localStorage.getItem(namespacedKey);
      if (
        currentLocal &&
        currentLocal !== 'undefined' &&
        currentLocal !== 'null'
      ) {
        const jsValue = JSON.parse(currentLocal);
        const woke = fromJS(jsValue);
        inMemoryCache = inMemoryCache.set(key, woke);
      }
    }
  } catch (err) {
    console.error(err);
    // if JSON.parse() fails, just blow this key away!
    set(key, undefined, false);
  }
  return inMemoryCache.get(key) || defaultValue;
}
