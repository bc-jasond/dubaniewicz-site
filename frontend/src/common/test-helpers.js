import { fromJS } from 'immutable';
import { reviver } from './utils';
import { SELECTION_LENGTH, SELECTION_NEXT } from './constants';

export function overrideConsole() {
  global.console = {
    info: jest.fn(),
    warn: jest.fn(),
    error: jest.fn()
  };
}

export function mockLocalStorage() {
  global.localStorageStorage = {};
  global.localStorage = {
    getItem: k => global.localStorageStorage[k],
    setItem: (k, v) => (global.localStorageStorage[k] = v),
    clear: () => (global.localStorageStorage = {})
  };
}

export function mockSetTimeout() {
  global.setTimeout = jest.fn();
}

export function makeSelections(values) {
  const head = {};
  let current = head;
  let prev;
  do {
    let [currentLength, ...currentValues] = values.shift();
    current[SELECTION_LENGTH] = currentLength;
    currentValues.forEach(v => {
      if (typeof v === 'object') {
        current[v.key] = v.value;
      } else {
        current[v] = true;
      }
    });
    if (prev) {
      prev[SELECTION_NEXT] = current;
    }
    prev = current;
    current = {};
  } while (values.length);
  prev[SELECTION_NEXT] = undefined; //important for the reviver()
  prev[SELECTION_LENGTH] = -1;
  return fromJS(head, reviver);
}
