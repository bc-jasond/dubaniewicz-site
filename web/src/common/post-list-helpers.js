import { getApiClientInstance } from './api-client';
import { formatPostDate } from './utils';
import { loading } from '../stores';

export function pushHistory(param, value) {
  const urlSearchParams = new URLSearchParams(window.location.search);
  if (value) {
    urlSearchParams.set(param, value === true ? '' : value);
  } else {
    urlSearchParams.delete(param);
  }
  const updatedQueryString =
    urlSearchParams.toString().length > 0
      ? `?${urlSearchParams.toString()}`
      : '';
  // update the URL in history for the user to retain, only if it changed!
  if (window.location.search != updatedQueryString) {
    window.history.pushState(
      {},
      document.title,
      `${window.location.pathname}${updatedQueryString}`
    );
  }
  return urlSearchParams;
}

export async function loadPosts(url, urlSearchParams, fetchClient) {
  loading.set(true);
  const apiClient = getApiClientInstance(fetchClient);
  const params = urlSearchParams.toString();
  const queryString = params.length > 0 ? `?${params}` : '';
  const { error, data: postsData } = await apiClient.get(
    `${url}${queryString}`
  );
  loading.set(false);
  return postsData.map((post) => ({
    ...post,
    published: formatPostDate(post.published),
    updated: formatPostDate(post.updated),
  }));
}
