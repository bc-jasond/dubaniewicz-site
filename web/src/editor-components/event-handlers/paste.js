import { KEYCODE_V } from '@filbert/constants';
import { getHighlightedSelectionOffsets } from '../../common/dom';
import { stopAndPrevent } from '../../common/utils';
import { doDelete } from '../editor-commands/delete';
import { doPaste } from '../editor-commands/paste';

let unexecuteSelectionOffsets;
// for a highlight-and-paste, this stores both the delete and paste history into one atomic unit
let pasteHistoryState = [];

// handle both the "paste" event and keyboard shortcut
export function isPasteEvent(evt) {
  return (
    evt.type === 'paste' ||
    ((evt.metaKey || evt.ctrlKey) && evt.keyCode === KEYCODE_V)
  );
}

export async function handlePaste({
  evt,
  selectionOffsets: selectionOffsetsArg,
  documentModel,
  historyManager,
  commitUpdates,
  closeAllEditContentMenus,
}) {
  // NOTE: this handler needs to pass through twice on a "paste" event
  // 1st: on "keydown" - this is to handle deleting selected text TODO: why? let the keydown noop
  // 2nd: on "paste" - now that the selection is clear, paste in the text
  if (!isPasteEvent(evt)) {
    return false;
  }

  const selectionOffsets =
    selectionOffsetsArg || getHighlightedSelectionOffsets();
  // if we're coming from "keydown" - check for a highlighted selection and delete it, then bail
  // we'll come back through from "paste" with clipboard data...
  if (evt.type !== 'paste') {
    // for undo (unexecute) history - need to store the "first" selections, aka before the delete operation
    unexecuteSelectionOffsets = selectionOffsets;
    if (selectionOffsets.caretStart !== selectionOffsets.caretEnd) {
      const { historyState } = doDelete(documentModel, selectionOffsets);
      pasteHistoryState.push(...historyState);
    }
    return true;
  }
  // NOTE: if these get called on the 'keydown' event, they'll cancel the 'paste' event
  stopAndPrevent(evt);

  const { selectionOffsets: executeSelectionOffsets, historyState } = doPaste(
    documentModel,
    selectionOffsets,
    evt.clipboardData
  );
  pasteHistoryState.push(...historyState);
  // TODO: paste into Meta Type nodes isn't supported
  // TODO: is this a valid state for this code anymore?
  if (!executeSelectionOffsets) {
    console.error('PASTE - executeSelectionOffsets is falsy?');
  }
  // add history entry
  historyManager.appendToHistoryLog({
    selectionOffsets: executeSelectionOffsets,
    historyState: pasteHistoryState,
  });

  pasteHistoryState = [];
  // for commitUpdates() -> setCaret()
  //await closeAllEditContentMenus();
  await commitUpdates(executeSelectionOffsets);
  return true;
}
