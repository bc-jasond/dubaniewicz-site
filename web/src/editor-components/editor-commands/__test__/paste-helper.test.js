import { overrideConsole } from '../../../common/test-helpers';
import DocumentModel from '@filbert/document/document-model';
import * as selectionHelpers from '@filbert/selection';
import { doPaste } from '../paste';
import {
  testPostWithAllTypesJS,
  imgId,
  firstNodeIdH1,
  firstNodeContent,
  firstPId,
} from '@filbert/util/test-post-with-all-types';

const { post, contentNodes } = testPostWithAllTypesJS;
overrideConsole();
let doc;

const mockText = 'some sample string';
const clipboardDataMock = {
  getData: () => mockText,
};

const spyAdjust = jest
  .spyOn(selectionHelpers, 'adjustSelectionOffsetsAndCleanup')
  .mockImplementation((...args) => args[0]);

beforeEach(() => {
  jest.clearAllMocks();
  doc = DocumentModel(post.id, contentNodes);
});
describe('Document Model -> paste node helper doPaste()', () => {
  test('validates input', () => {
    expect(() => {
      // bad node id
      doPaste(doc, {}, clipboardDataMock);
    }).toThrow();
    expect(() => {
      // meta type nodes not supported
      doPaste(doc, { caretStart: 0, startNodeId: imgId }, clipboardDataMock);
    });
    expect(spyAdjust).not.toHaveBeenCalled();
  });
  test('pastes a single line of text', () => {
    const caretStartArg = 10;
    const {
      historyState,
      executeSelectionOffsets: { startNodeId, caretStart },
    } = doPaste(
      doc,
      { caretStart: caretStartArg, startNodeId: firstNodeIdH1 },
      clipboardDataMock
    );
    expect(startNodeId).toEqual(firstNodeIdH1);
    expect(caretStart).toEqual(caretStartArg + mockText.length);
    expect(historyState).toMatchSnapshot();
    expect(spyAdjust).toHaveBeenCalledWith(
      doc.getNode(firstNodeIdH1),
      firstNodeContent,
      caretStartArg,
      mockText.length
    );
  });
  test('pastes plain text - multi line paste', () => {
    const clipboardText = 'bib jibbs\nbab jabs\nbubba rubba jubbson';
    const textLines = clipboardText.split('\n');
    const lastLine = textLines[textLines.length - 1];
    const clipboardDataMockMultiLine = {
      getData: () => clipboardText,
    };
    const caretStartArg = Math.floor(firstNodeIdH1.length / 2);
    const originalType = doc.getNode(firstNodeIdH1).get('type');

    const {
      historyState,
      executeSelectionOffsets: { startNodeId, caretStart },
    } = doPaste(
      doc,
      { startNodeId: firstNodeIdH1, caretStart: caretStartArg },
      clipboardDataMockMultiLine
    );
    expect(startNodeId).not.toBe(firstNodeIdH1);
    expect(caretStart).toBe(lastLine.length);
    expect(historyState.length).toBe(6);
    expect(historyState.pop().executeState.get('next_sibling_id')).toBe(
      firstPId
    );
    expect(spyAdjust).toHaveBeenCalledTimes(2);
    let current = doc.getNode(firstNodeIdH1);
    for (let i = 0; i < textLines.length; i++) {
      expect(current.get('type')).toBe(originalType);
      current = doc.getNextNode(current.get('id'));
    }
  });
});
