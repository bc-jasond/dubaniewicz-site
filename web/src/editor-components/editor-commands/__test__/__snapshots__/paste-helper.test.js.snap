// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document Model -> paste node helper doPaste() pastes a single line of text 1`] = `
Array [
  Object {
    "executeState": Immutable.Map {
      "post_id": 175,
      "id": "8e34",
      "next_sibling_id": "621e",
      "type": "h1",
      "content": "Large Headsome sample stringing",
      "meta": Immutable.Map {},
    },
    "unexecuteState": Immutable.Map {
      "post_id": 175,
      "id": "8e34",
      "next_sibling_id": "621e",
      "type": "h1",
      "content": "Large Heading",
      "meta": Immutable.Map {},
    },
  },
]
`;
